### DISCUSSION TITLE
ReaLearn OSC: How to Send Track Colors to TouchOSC on Project Load or Polling?

### DISCUSSION THREAD

**Discussion by TheSoundOfOneHandClapping** [helgoboss/helgobox#1133] [Q&A] [answered]
I've set up ReaLearn to send track colors to TouchOSC. I've found the setup pictured below to work. However, it only sends the color when the track is manually selected.

What would be preferable is if the colors of all tracks were sent on project load, or even better, if TouchOSC could poll the colors of specific tracks. What is the best approach to achieve this?

<img width="716" alt="Screenshot 2024-08-26 at 16 20 26" src="https://github.com/user-attachments/assets/2acee915-2f5a-4f12-952c-df4c23d68236">

**Comment by helgoboss** [comment]
Are you absolutely sure that it doesn't send on project load? Have you checked the real-target output logs (see Menu)

**Reply by TheSoundOfOneHandClapping** [reply]
Sorry, I probably wasn‘t clear enough in my description.
Yes, I’ve checked, ReaLearn sends the color of only **one** track on project load.
But as TouchOSC displays several tracks simultaneously, I‘d like to color them **all** on project load.
To clarify my intentions, following would be a brute force solution to my problem: On Project Load, let TouchOSC go through all tracks, select one track after the other and use the colors that ReaLearn is returning to color the tracks in TouchOSC. This would be obviously very clumsy.
But I bet ReaLearn has a much more elegant solution for this?

**Reply by helgoboss** [reply]
Ah, now I understand. No, polling would be quite inefficient.

The problem is in your configuration. You have just one mapping and it refers to the currently selected track. And this is exactly what you get: The color of the currently selected track - no more, no less. If you want to get the color for each track, you need one mapping per track, each one using a different OSC address as source and a different track number ("Positioned") as target. Tracks which don't exist, will automatically return a "none" color (I guess "black", not sure now). ReaLearn will push color changes to TouchOSC whenever it's necessary: On project load, on track changes/additions/removals etc.

Building these via GUI is not a good idea, it will be too much clicking and repetition. Since you use TouchOSC, you are probably familiar with Lua. You can use Lua (or rather Luau) in ReaLearn as well, to build those mappings. You can start by building one or two mappings in the GUI and check if it works. When you are satisfied, migrate to Lua: Press "Export to clipboard" and "Main compartment as Lua". You will get a valid Lua/Luau table that represents those mappings. Now you can use your Lua skills (for loops etc.) to create lots of mappings with just a few lines of code. Then copy that and press "Import from clipboard".

**Reply by TheSoundOfOneHandClapping** [reply]
Thanks, we‘re on the right track.
Only thing is this: you write „each one using a different OSC address as source“ - isn‘t ist possible to use the same OSC address for every track and pass the track number as parameter? (I tried but I couldn‘t get it to work)
If we have a different OSC address for every track, we have to create one receiving element per track, which is a lot of work and not very elegant.

**Reply by helgoboss** [reply]
Uhm, yes, you could enter the same OSC address in each mapping. Then it would send lots of color values in sequence. But how do you know which track is which? I mean, how do you imagine this to work on a purely conceptual level?

Also, how will your TouchOSC layout roughly look? On the one hand, you say you need the colors of all tracks ... which suggests that you want to display multiple tracks separately (and in which case it would probably make sense to use different addresses, e.g. "track/1/color", "track/2/color"). On the other hand, you dislike the idea of using different addresses, which suggests you don't want to display different tracks. Could you talk a bit more about your goal?

**Reply by helgoboss** [reply]
Ah, sorry. I missed that "pass the track number as parameter" suggestions. So you can ignore the question "But how do you know which track is which?". Still interested in the answer to the other questions though.

**Reply by TheSoundOfOneHandClapping** [reply]
Thank you for your time and patience.

First part: How would I handle the incoming OSC message, that had two parameters: color and track number? Like this: I introduce a (hidden) element that receives the OSC message and in the onReceiveOSC() callback I read out the parameters and set the corresponding track number to the specified color. No problem to do that with Touch OSC whatsoever.

Second part:
My OSC layout is based on the native Reaper OSC implementation, which bundles tracks into sizable chunks (called banks). And there‘s actually a big difference between how Reaper handles tracks for OSC and how ReaLearn handles it. I try to explain:

The native Reaper OSC implementation is configured in default.ReaperOSC. And there, the first thing you can define is „DEVICE_TRACK_COUNT“, which is how many tracks this device can display at once (the track bank size) With this info, Reaper OSC bundles the tracks in a project into chunks the size of DEVICE_TRACK_COUNT, it increments a parameter called „bank“ when crossing a DEVICE_TRACK_COUNT border, and the track count starts from 1 again.

Example, let’s say the DEVICE_TRACK_COUNT is 8, Reaper would sends e.g.
„track/1/mute“ when Track 1 is selected, but it also sends the **exact same** OSC message when Track 9 or 17 or 25 etc is selected. The parameter „bank“ is then 1 for Track 1, 2 for Track 9, 3 for Track 17, etc.

So on the TouchOSC-side you only have to handle tracks 1  through 8 plus a bank parameter. You will never receive a OSC message outside of that range, as the bank parameter handles that.

In ReaLearn on the other hand, this „bank“-mechanism is missing (or not obvious?) and I don‘t think tracks from 1 to virtually infinity can be handled in TouchOSC in a reasonable way.

I‘ve seen some other questions pop up where people wanted to build this „bank“ system with ReaLearn but failed.

Third part:
When I use the same OSC address in each mapping, ReaLearn sends an **endless** stream of color values in sequence without ever stopping. What could I be doing wrong?

**Reply by helgoboss** [reply, marked as answer]
> Thank you for your time and patience.
> 
> First part: How would I handle the incoming OSC message, that had two parameters: color and track number? Like this: I introduce a (hidden) element that receives the OSC message and in the onReceiveOSC() callback I read out the parameters and set the corresponding track number to the specified color. No problem to do that with Touch OSC whatsoever.

Did you want to write how to *create* multi-argument OSC messages in ReaLearn (not *handle* them)? Looks like you already described how you handle them in TouchOSC. 

> Second part: My OSC layout is based on the native Reaper OSC implementation, which bundles tracks into sizable chunks (called banks). And there‘s actually a big difference between how Reaper handles tracks for OSC and how ReaLearn handles it. I try to explain:
> 
> The native Reaper OSC implementation is configured in default.ReaperOSC. And there, the first thing you can define is „DEVICE_TRACK_COUNT“, which is how many tracks this device can display at once (the track bank size) With this info, Reaper OSC bundles the tracks in a project into chunks the size of DEVICE_TRACK_COUNT, it increments a parameter called „bank“ when crossing a DEVICE_TRACK_COUNT border, and the track count starts from 1 again.
> 
> Example, let’s say the DEVICE_TRACK_COUNT is 8, Reaper would sends e.g. „track/1/mute“ when Track 1 is selected, but it also sends the **exact same** OSC message when Track 9 or 17 or 25 etc is selected. The parameter „bank“ is then 1 for Track 1, 2 for Track 9, 3 for Track 17, etc.
> 
> So on the TouchOSC-side you only have to handle tracks 1 through 8 plus a bank parameter. You will never receive a OSC message outside of that range, as the bank parameter handles that.
> 
> In ReaLearn on the other hand, this „bank“-mechanism is missing (or not obvious?) and I don‘t think tracks from 1 to virtually infinity can be handled in TouchOSC in a reasonable way.
> 
> I‘ve seen some other questions pop up where people wanted to build this „bank“ system with ReaLearn but failed.

ReaLearn has dynamic track selectors (`<Dynamic>`) and conditional activation. With this, you can emulate bank systems and other stuff. This gives you the freedom but also responsibility to create those "chunks" yourself. I don't have the time to explain this in detail now, but the puzzle pieces can be found in the technical reference. I plan to make a tutorial about this. You can see it in action in the "DAW control" preset, which defaults to banks of 8 and defines the currently selected track (variable `selected_track_index`) as the first track in such a bank. You could also use a compartment parameter to control which bank is active, but that's a bit more involved.

> Third part: When I use the same OSC address in each mapping, ReaLearn sends an **endless** stream of color values in sequence without ever stopping. What could I be doing wrong?

You can send the preset at info@helgoboss.org and I can have a look.

**Reply by helgoboss** [reply]
Still a few hints regarding the banking (assuming banks of 8):

- You need at least 8 mappings for transmitting the colors:
    1. **Source:** OSC address "track/1/color" | **Target:** "Track" with `<Dynamic>` expression `selected_track_index`
    2. **Source:** OSC address "track/2/color" | **Target:** "Track" with `<Dynamic>` expression `selected_track_index + 1`
    3. **Source:** OSC address "track/2/color" | **Target:** "Track" with `<Dynamic>` expression `selected_track_index + 2`
    4. and so on

**Reply by TheSoundOfOneHandClapping** [reply]
Thanks for all the info.
Sent you am email with the preset in question. I add it here as well if someone else is interested.
[track-color-test.json](https://github.com/user-attachments/files/16759633/track-color-test.json)
The preset uses the same OSC address in each mapping (which works nicely) but the problem is, ReaLearn sends an endless stream of messages in sequence without ever stopping.

**Reply by helgoboss** [reply]
I can investigate that. But apart from that, why don't you use different addresses? With track banking in place, all you need is 8 addresses. Still don't know how your layout is actually supposed to look. A simple screenshot would help a great deal.

**Reply by helgoboss** [reply]
The more I think about this, the less I can see the "single address" method work nicely. ReaLearn has mechanisms for detecting when a particular source (LED, OSC label, whatever) is not in use anymore. When a target deactivates (e.g. a track disappears) and ReaLearn detects that the corresponding source is not in use anymore in other mappings, it switches it off (e.g. sends a value of 0 or an empty string or color back). This mechanism can't work if you use multiple mappings with the same OSC address at the same time. So my guess is that you will not get the desired behavior when removing a track. And maybe other issues.

**Reply by TheSoundOfOneHandClapping** [reply]
> But apart from that, why don't you use different addresses? With track banking in place, all you need is 8 addresses.

The TouchOSC project is based on Reaper‘s native OSC implementation, ReaLearn is only used to augment information that is not or only unreliably available from Reaper‘s OSC.
As the project is quite complex and very well debugged, it would take me a ton of time to rewrite everything to work exclusively with ReaLearn.

Yeah, you‘re probably right, let‘s forget about it.
Thanks!

**Reply by helgoboss** [reply]
I will have a look when there's some time.

I'm not sure if that would work though. When REAPER opens and UDP port and ReaLearn the same one, it will clash. One of them won't work. Did you take that into account?

UPDATE: Ah, just reading that you already abandoned the idea. Yeah, I think it wouldn't work to combine both REAPER OSC and ReaLearn OSC because of the UDP port conflict. And even if that worked somehow, it would become a monster. So many moving parts. When it comes to OSC, I'm afraid it's an either-or: ReaLearn or REAPER or CSI. They don't really go together AFAIK. With MIDI it's different. REAPER and ReaLearn can share the same devices.

**Reply by TheSoundOfOneHandClapping** [reply]
> When REAPER opens and UDP port and ReaLearn the same one, it will clash.

Not true. As I said before, ReaperOSC & ReaLearnOSC are working nicely together - I use it everyday without problem.
Just the tracks are not colored, which this question was all about.
You rack your brains over things that have already been solved. :-)

**Reply by helgoboss** [reply]
How did you solve it? Or did it just work, the same UDP ports not conflicting with each other?

**Reply by TheSoundOfOneHandClapping** [reply]
It’s two separate ports, I can post a screenshot of the settings when I‘m back in the studio.

**Reply by helgoboss** [reply]
Ah, you are using two pairs of ports in TouchOSC. Okay, that makes sense.

**Reply by TheSoundOfOneHandClapping** [reply]
Ok, I managed to make it work with a compartment parameter acting as bank number and multiple OSC messages. Case closed.
Thank you for the guidance.

**Comment by helgoboss** [comment]
Just for the logs, here's why the endless stream of feedback happens in above case:

> ReaLearn has a duplicate-feedback detection mechanism that is important for feedback that is polled regularly (such as track colors). Under normal circumstances, it prevents that endless stream.

> However, in your case, the first mapping sends feedback "1, color1" to "ReaLearn/Track/Color/" and the second "2, color2". Which is 2 different feedback messages, not duplicates.

Possible solution, but I guess we'll never need that:

> I guess for such cases, the logically correct solution would be to be able to declare an OSC argument as part of the source address. Then "ReaLearn/Track/Color/" with the first argument being "1" would be considered as a different source than "ReaLearn/Track/Color" with the first argument being "2". However, such a mechanism doesn't currently exist.
